internal/graph/lookupsubjects.go

```diff
@@ -342,47 +351,53 @@ type lookupSubjectsReducer interface {
 // Union
 type lookupSubjectsUnion struct {
 	parentStream dispatch.LookupSubjectsStream
-	encountered  *util.Set[string]
-	mu           sync.Mutex
+	collectors   map[int]*dispatch.CollectingDispatchStream[*v1.DispatchLookupSubjectsResponse]
 }
 
 func newLookupSubjectsUnion(parentStream dispatch.LookupSubjectsStream) *lookupSubjectsUnion {
 	return &lookupSubjectsUnion{
 		parentStream: parentStream,
-		encountered:  util.NewSet[string](),
-		mu:           sync.Mutex{},
+		collectors:   map[int]*dispatch.CollectingDispatchStream[*v1.DispatchLookupSubjectsResponse]{},
 	}
 }
 
 func (lsu *lookupSubjectsUnion) ForIndex(ctx context.Context, setOperationIndex int) dispatch.LookupSubjectsStream {
-	return &dispatch.WrappedDispatchStream[*v1.DispatchLookupSubjectsResponse]{
-		Stream: lsu.parentStream,
-		Ctx:    ctx,
-		Processor: func(result *v1.DispatchLookupSubjectsResponse) (*v1.DispatchLookupSubjectsResponse, bool, error) {
-			lsu.mu.Lock()
-			defer lsu.mu.Unlock()
-
-			filtered := make([]string, 0, len(result.FoundSubjectIds))
-			for _, subjectID := range result.FoundSubjectIds {
-				if lsu.encountered.Add(subjectID) {
-					filtered = append(filtered, subjectID)
-				}
-			}
+	collector := dispatch.NewCollectingDispatchStream[*v1.DispatchLookupSubjectsResponse](ctx)
+	lsu.collectors[setOperationIndex] = collector
+	return collector
+}
 
-			if len(filtered) == 0 {
-				return nil, false, nil
-			}
+func (lsu *lookupSubjectsUnion) CompletedChildOperations() error {
+	var foundSubjects util.SubjectSet
+	metadata := emptyMetadata
 
-			return &v1.DispatchLookupSubjectsResponse{
-				FoundSubjectIds: filtered,
-				Metadata:        result.Metadata,
-			}, true, nil
-		},
+	for index := 0; index < len(lsu.collectors); index++ {
+		collector, ok := lsu.collectors[index]
+		if !ok {
+			return fmt.Errorf("missing collector for index %d", index)
+		}
+
+		results := util.NewSubjectSet()
+		for _, result := range collector.Results() {
+			metadata = combineResponseMetadata(metadata, result.Metadata)
+			results.UnionWith(result.FoundSubjects)
+		}
+
+		if index == 0 {
+			foundSubjects = results
+		} else {
+			foundSubjects.UnionWithSet(results)
+		}
```

Comment:
Does this make a huge perf difference that you shouldn't just always Union and preallocate `foundSubjects := util.NewSubjectSet()` before the loop?

Reply:
Nope; this was mostly a copy-paste thing. I can change it